### PR TITLE
Added Binary Search [Python]

### LINKED ISSUE #59: Binary Search [python]

### PR DESCRIPTION
**Pull request by aashutoshrathi**
Fixed #59 

#### By submitting this pull request I confirm I've read and complied with the below requirements.

- [x] I have read the [Contribution guidelines](https://github.com/iiitv/algos/blob/master/CONTRIBUTING.md) and I am confident that my PR reflects them.
- [x] I have followed the [coding guidelines](https://github.com/iiitv/algos/blob/master/CONTRIBUTING.md#cs) for this project.
- [x] My code follows the [skeleton code structure](https://github.com/iiitv/algos/blob/master/CONTRIBUTING.md#sample).
- [x] This pull request has a descriptive title. For example, `Added {Algorithm/DS name} [{Language}]`, not `Update README.md` or `Added new code`.

### CODE REVIEW

**Comment by aashutoshrathi** [comment]
@aviaryan @Monal5031 @singhpratyush What should I use instead of randint() so that "Issue found: Standard pseudo-random generators are not suitable for security/cryptographic purposes.", gets fixed 😕

**Comment by aviaryan** [comment]
@aashutoshrathi Removed that test from codacy now. So no worries.

**Comment by aashutoshrathi** [comment]
Should I change randint() to something else ?

**Comment by aashutoshrathi** [comment]
@Monal5031 check for indentations now

**Comment by Monal5031** [comment]
@aashutoshrathi Indentations look fine but the line number 6 and 36 seem to cross the 75 chars per line limit.
Try shortening the comment or you can add it separate line

**Comment by aashutoshrathi** [comment]
@Monal5031 Done

File: binary_search/binary_search.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
+from random import randint
+
+
+def binary_search_iterative(array, left, right, element):
```
Comment (singhpratyush) [inline binary_search/binary_search.py]: * Need only two params here - `array` and `element`. Others can be generated.
* Function names should be `binary_search` and `binary_search_recursive`.

File: binary_search/binary_search.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    while left <= right:
+        mid = int(left + (right - left) / 2)
```
Comment (singhpratyush) [inline binary_search/binary_search.py]: How about
```python
    mid = int((left + right) / 2)
```

File: binary_search/binary_search.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        # indices of a list must be integer
+        if array[mid] == element:
+            return mid
+        elif array[mid] < element:
+            left = mid + 1
+        else:
+            right = mid - 1
+    return -1
```
Comment (singhpratyush) [inline binary_search/binary_search.py]: Index `-1` means something in Python. Please be careful here.

File: binary_search/binary_search.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+
+def binary_search_recursive(array, left, right, element):
```
Comment (singhpratyush) [inline binary_search/binary_search.py]: Can set default values to `left` and `right` as `0` and `None` respectively. If `None`, `right` can be set to `len(array)`.
```python
def binary_search_recursive(array, element, left=0, right=None):
    right = len(array) if right is None else right
    ...
```
How would it be?

File: binary_search/binary_search.py (hunk 2 above)
Comment (aashutoshrathi) [inline binary_search/binary_search.py]: done 👍

File: binary_search/binary_search.py (hunk 4 above)
Comment (Monal5031) [inline binary_search/binary_search.py]: @singhpratyush if we set left =0 in each recursion would that give correct answer?
suppose our element is in the right half then we set left = mid+1, wouldnt that create a problem?

Comment (aashutoshrathi) [inline binary_search/binary_search.py]: unable to do this particular change, take a look at remaining file

Comment (singhpratyush) [inline binary_search/binary_search.py]: @Monal5031: That is the default value for the parameter. This example may help - 
```python
def func(a='User'):
    print('Hello ' + a)

func()  # Hello User
func('Pratyush')  # Hello Pratyush
```
@aashutoshrathi : Please reach out if any issues.

Comment (Monal5031) [inline binary_search/binary_search.py]: @singhpratyush Thanks for the information, learnt something new 😄

File: binary_search/binary_search.py (hunk 3 above)
Comment (aashutoshrathi) [inline binary_search/binary_search.py]: replaced it with None

File: binary_search/binary_search.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+        if array[mid] == element:
+            return mid
+        elif array[mid] < element:
+            left = mid + 1
+        else:
+            right = mid - 1
+    return None
+
+
+def binary_search_recursive(array, element, left = 0, right = None):
```
Comment (singhpratyush) [inline binary_search/binary_search.py]: No space around `=`.
```python
def binary_search_recursive(array, element, left=0, right=None):
```
Also, if the default value for right is `None`, your code will fail if user doesn't passes right as an argument.

File: binary_search/binary_search.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    domain = 100  # user can change it
+    array = [random.randint(0, domain) for i in range(size)]
+    # print ("Array :", array)
+    array.sort(key=int)
+    # print ("Sorted Array :", array)
+    element = random.randint(0, domain)
+    # i.e. a random element can be selected from domain
+    print('Element = ', element)
+    length = len(array)
+    result = binary_search_recursive(array, element, 0, length - 1)
```
Comment (singhpratyush) [inline binary_search/binary_search.py]: The whole idea behind using default values for parameters was to ease out the things for the user. If user wishes to use your function to perform binary search using recursion, right now he would be calling
```python
index = binary_search_recursive(array, element, 0, length - 1)
```
But if you utilize power of default param values, you could provide user with an option that he would pass only array and element and by default you will search the whole array -
```python
index = binary_search_recursive(array, element)
```

File: binary_search/binary_search.py (hunk 5 above)
Comment (aashutoshrathi) [inline binary_search/binary_search.py]: No, it's somehow working 😕

**Comment by mohitkyadav** [comment]
git

I can see element 10 is at index 4.

**Comment by Monal5031** [comment]
@mohitkyadav Binary Search is for sorted array's only

**Comment by mohitkyadav** [comment]
@Monal5031,  I know but I just changed the value of domain in the code, that's it...... Try to run yourself.
The input was self-generated.

**Comment by mohitkyadav** [comment]
screenshot 22

**Comment by Monal5031** [comment]
@mohitkyadav you don't have to delete my comments 😜

File: binary_search/binary_search.py (hunk 6 above)
Comment (aashutoshrathi) [inline binary_search/binary_search.py]: this changes has been done, take a look now @singhpratyush

**Comment by singhpratyush** [comment]
Why are the indices different if the algorithm is meant to give same results? Please justify that this should happen else make proper changes.

**Comment by aashutoshrathi** [comment]
@singhpratyush when there are repeating numbers, recursive is returning index of second last occurrence of element and iterative is returning index of last occurrence.
Should I change that ?

**Comment by singhpratyush** [comment]
Indeed they are giving different results. Since both the methods are supposed to progress in the same way, results should be identical (according to my logic). @aashutoshrathi: Please correct me if wrong.

**Comment by Monal5031** [comment]
@singhpratyush @aashutoshrathi If you look closely to the solutions particularly the lines 19 and 41 you would get why the semantic error is occuring.

**Comment by aashutoshrathi** [comment]
@Monal5031 @singhpratyush actually in iterative `right = len(array) - 1` , while it should be `right = len(array)`, 
I fixed it take a look 👍

**Comment by singhpratyush** [comment]
@aviaryan @rajeeviiit: Take a look.

**Pull request merged**